### ISSUE TITLE
False positive in dead code because checker receives no flow info

### ISSUE THREAD

**Issue by cpovirk** [typetools/checker-framework#4265]
\[edit: I've updated the issue title to reflect Werner's explanation of what's going on. My report below may make more sense if read with my initial title, which was "Assignments to local variables inside a `catch` have no effect, even inside that `catch`."\]

For example, if I assign a non-null value to a variable inside a `catch`, I still get an error when I subsequently dereference it inside the `catch`. I would have expected the dereference to be safe, as it is if I perform the process outside a `catch`.

```
$ cat AssignInCatch.java 
import org.checkerframework.checker.nullness.qual.Nullable;

class AssignInCatch {
  void foo(@Nullable Object o) {
    try {
    } catch (Exception e) {
      o = "";
      o.toString();
    }
  }
}

$ checker/bin/javac -processor org.checkerframework.checker.nullness.NullnessChecker AssignInCatch.java
AssignInCatch.java:8: error: [dereference.of.nullable] dereference of possibly-null reference o
      o.toString();
      ^
1 error
```

That's using version 3.10.0.

**Comment by wmdietl** [comment]
Related: https://github.com/typetools/checker-framework/issues/4018#issuecomment-744861768
As the try block is empty, there currently is no dataflow edge to any of the catch blocks. Dataflow refinement only happens for live blocks.

**Comment by cpovirk** [comment]
Ah, thanks! I was getting very confused.

As you would expect, this did not happen in "real code," so I am not particularly concerned about it. I just over-minimized my test case.

**Comment by mernst** [comment]
The fact that checkers can issue false positive warnings in dead code is documented in the penultimate bullet point of [the "Unexpected type-checking results"](https://checkerframework.org/manual/#common-problems-typechecking) section of the manual.  (Changing that behavior would be possible, but does not seem like a priority.)
